### PR TITLE
Show Quickshell panels on all active monitors

### PR DESCRIPTION
**Pull request by ChrisTitusTech**
## Summary

- create one managed `DwmPanel` per active `Quickshell.screens` entry
- keep shared state, audio, network, Bluetooth, tray, and control-center providers singleton
- anchor panel-triggered popups to the monitor whose bar was clicked
- keep unsolicited notifications on the default panel
- publish and reconcile per-monitor desktop ownership through `_DWM_MONITOR_DESKTOPS`
- harden panel, client, focus, and EWMH state across monitor geometry changes

## Root cause

The shell instantiated one `DwmPanel` without assigning per-screen instances, so Quickshell placed the bar only on its default screen even when X11 exposed multiple active monitors. The resulting multi-monitor implementation also required explicit dwm reconciliation for tag ownership, panel registration, client migration, and focus during hotplug transitions.

## Automated validation

Final head: `726a823f9fac91c948d694fdb98f52507263412b`

- `make clean` - PASS
- `make` - PASS
- `make check` - PASS, including clean C build, shell/static checks, Quickshell QML and nested-X11 tests, installers, Kickstart checks, staged installation, and reproducible release validation
- `sh tests/test-monitor-tag-switching.sh` - PASS
- `shellcheck tests/test-monitor-tag-switching.sh` - PASS
- `shfmt -d tests/test-monitor-tag-switching.sh` - PASS
- `git diff --check` - PASS
- GitHub CI `validate`, `clang-build`, Debian/Arch/RHEL smoke tests, and `quickshell-qml` - PASS
- CodeQL `analyze` and `CodeQL` - PASS
- documentation `build` - PASS; deploy correctly skipped for the PR
- remote CodeRabbit review on the final head - PASS

The local CodeRabbit CLI review attempt for the uncommitted final fix was temporarily rate-limited. The remote CodeRabbit review completed against the pushed final head with no new actionable thread.

## Review evidence

Review findings fixed across the branch include disconnected-panel fallback, monitor/tag state reconciliation, panel height and geometry handling, stale EWMH client removal, surviving-panel rebind, vertical dock rejection, hidden-panel preservation, configure-event query cost, and selected-client focus preservation.

The fresh Codex review on `726a823` raised one final P1 concern about a dangling source-monitor selection. It was verified as not applicable: `detachstack(c)` replaces the source monitor selection before `c->mon` is assigned to the destination, after which destination selection and active focus are restored. The thread was answered with that ordering evidence and resolved without redundant code.

All review threads are resolved. Fresh CodeRabbit and Codex reviews both completed against the final head.

## Manual validation

Environment: Fedora X11/dwm, Quickshell 0.2.1, two active 2560x1440 monitors.

The user-visible Quickshell behavior was validated before the final C-only review fix:

- `xrandr --listactivemonitors` reported two active outputs: `DVI-I-1-1` at `0,0` and primary `DVI-I-2-2` at `2560,0`
- exactly two visible Quickshell panel windows were present, each `2560x30`, at `0,0` and `2560,0`
- clicking the left bar opened Control Center on the left monitor (`x=0`)
- clicking the right bar opened Control Center on the right monitor (`x=2560`)
- the managed installed QML tree matched the repository tree
- one Quickshell process was running, with no orphaned `dwm-quickshell-*` helpers and no blocking QML log errors
- three consecutive two-panel closed-idle samples measured 3.00%, 2.90%, and 3.00% of one CPU core
- dwm remained running throughout; only the managed Quickshell process was restarted

A local 5120x80 screenshot was inspected to confirm both rendered bars; it is not committed as a generated repository asset.

## Known limitation

Real hardware validation covered two side-by-side 2560x1440 X11 outputs. Dynamic physical unplug/replug was not performed. The final hotplug edge cases are covered by focused source assertions, clean GCC and Clang builds, the repository test suite, nested-X11 validation, and fresh independent review, but not by a physical unplug/replug test on the final binary.

### CODE REVIEW

**Comment by ChrisTitusTech** [comment]
@coderabbitai review

File: config/quickshell/shell.qml [hunk 1]
```diff
@@ -20,6 +20,14 @@ pragma ComponentBehavior: Bound
 ShellRoot {
     id: root
 
+    property var selectedPanelWindow: null
+    readonly property var defaultPanelWindow: panelVariants.instances.length > 0
+        ? panelVariants.instances[0] : null
+    readonly property var activePanelWindow: selectedPanelWindow || defaultPanelWindow
+    readonly property var activePanelScreen: activePanelWindow
+        ? activePanelWindow.screen
+        : (Quickshell.screens.length > 0 ? Quickshell.screens[0] : null)
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply config/quickshell/shell.qml:30]: Fixed in 25c1bb8. `activePanelWindow` now accepts the selected panel only while `selectedPanelWindow.screen` is live, and `activePanelScreen` independently requires a live screen before falling back to `Quickshell.screens[0]`. Added a focused wiring assertion. `make check-quickshell-qml`, Control Center, System Health Xvfb, Settings Xvfb, ShellCheck, shfmt, and the complete `make check` all pass; final `make check` exited 0.

**Comment by ChrisTitusTech** [comment]
@coderabbitai review

**Comment by ChrisTitusTech** [comment]
@codex review

File: config/quickshell/shell.qml [hunk 2]
```diff
@@ -356,29 +365,37 @@ ShellRoot {
...
-        bluetoothModel: bluetoothModel
-        controlCenterModel: controlCenterModel
-        powerMenuModel: powerMenuModel
+        model: Quickshell.screens
+
+        DwmPanel {
+            required property var modelData
+
+            screen: modelData
+            state: dwmState
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply config/quickshell/shell.qml:393]: Fixed in 14346fa. The shared `DwmState` now consumes the event-driven `_DWM_MONITOR_DESKTOPS` root property, whose signed geometry/current-desktop tuples use the same logical order as `getmontagmask()`. Each `DwmPanel` derives its owned workspace indexes from its screen, displays only those tags, resolves that monitor's current tag, and rejects switches outside its ownership range. Nested X11 verifies property publication/updates; live validation showed tags 5-9 on the left bar and 1-4 on the right with independent selection. `make check-monitor-tags`, QML checks, and full `make check` pass.

File: config/quickshell/shell.qml [hunk 3]
```diff
@@ -356,29 +365,37 @@ ShellRoot {
...
+            clock: clock
+            networkModel: networkModel
+            controlsModel: controlsModel
+            bluetoothModel: bluetoothModel
+            controlCenterModel: controlCenterModel
+            powerMenuModel: powerMenuModel
+            onPopupRequested: panel => root.selectedPanelWindow = panel
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply config/quickshell/shell.qml]: Fixed in 14346fa. `selectPanelPopup(panel, popupId)` now closes every competing panel popup model before changing the shared anchor, while preserving the target model so its existing toggle behavior remains intact. Live two-monitor validation opened Control Center on the left, then a different panel popup on the right, and observed exactly one full-screen click-away surface anchored at x=2560. Focused QML/control-center checks and the complete `make check` pass.

File: config/quickshell/shell.qml [hunk 4]
```diff
@@ -387,18 +404,18 @@ ShellRoot {
...
         bluetoothModel: bluetoothModel
-        panelWindow: panelWindow
+        panelWindow: root.activePanelWindow
     }
 
     ControlCenterWindow {
         controlCenterModel: controlCenterModel
         launcherModel: launcherModel
-        panelWindow: panelWindow
+        panelWindow: root.activePanelWindow
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply config/quickshell/shell.qml:432]: Fixed in 14346fa. Control Center now captures its panel screen before closing and passes it through `LauncherModel`, `SettingsModel`, and `ControlCenterModel` utility navigation. `LauncherWindow`, `SettingsWindow`, and `UtilityDetailWindow` bind to those stored target screens; System Health also uses the captured value. The target remains stable if another bar is clicked afterward. Focused wiring assertions, QML lint, nested Settings/System Health tests, and the complete `make check` pass.

**Comment by ChrisTitusTech** [comment]
@codex review

File: dwm.c [hunk 5]
```diff
@@ -1018,6 +1019,7 @@ configurenotify(XEvent *e)
 			}
 			arrange(NULL);
 			focus(NULL);
+			updatecurrentdesktop();
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply dwm.c:1031]: Fixed in 33aee1e. The root ConfigureNotify path now calls `reconcilemonitortags()` immediately after `updategeom()` and before `updatecurrentdesktop()`. Reconciliation refreshes `monitorcount`, masks both tagsets to each monitor's current logical ownership, assigns a valid fallback when needed, synchronizes current/previous per-tag indexes, and restores the selected tag's saved layout/master/factor/bar state before arrange. The focused monitor-tag guard, clean C build, local CodeRabbit review, and complete `make check` all pass.

File: config/quickshell/shell.qml [hunk 6]
```diff
@@ -356,29 +378,37 @@ ShellRoot {
...
-        state: dwmState
-        clock: clock
-        networkModel: networkModel
-        controlsModel: controlsModel
-        bluetoothModel: bluetoothModel
-        controlCenterModel: controlCenterModel
-        powerMenuModel: powerMenuModel
+        model: Quickshell.screens
+
+        DwmPanel {
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply config/quickshell/shell.qml:389]: Fixed in 33aee1e. Each `DwmPanel` now receives an explicit `primaryPanel` flag from the first/default Quickshell screen, and a Loader instantiates `TrayArea` only when that flag is true. Secondary panels no longer construct or render the singleton tray collection. The focused tray guard, QML validation, local CodeRabbit review, live two-screen managed-shell restart, and complete `make check` all pass.

**Comment by ChrisTitusTech** [comment]
@codex review

**Comment by ChrisTitusTech** [comment]
@coderabbitai The outside-diff `_NET_CURRENT_DESKTOP` finding on 33aee1e is already satisfied in the current head. `updatecurrentdesktop()` computes the selected tag and calls `ewmh_replace_root_cardinal(netatom[NetCurrentDesktop], data, 1)` before constructing and publishing `_DWM_MONITOR_DESKTOPS`; the existing callers therefore update both properties together. `tests/test-xvfb-runtime.sh` also verifies `_NET_CURRENT_DESKTOP` changes from 0 to 1 after the nested tag switch, and the final `make check` passed. No code change is needed for this false positive.

File: config/quickshell/state/DwmState.qml [hunk 7]
```diff
@@ -59,6 +74,71 @@ Scope {
         return root.occupiedWorkspaces.indexOf(index) !== -1;
     }
 
+    function screenIndex(screen) {
+        for (let index = 0; index < root.monitorWorkspaceRows.length; index++) {
+            const row = root.monitorWorkspaceRows[index];
+
+            if (screen && row.x === screen.x && row.y === screen.y
+                    && row.width === screen.width && row.height === screen.height) {
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply config/quickshell/state/DwmState.qml]: Fixed in 99f9f70. `DwmState.screenIndex()` now derives device-pixel geometry using `screen.devicePixelRatio` and matches each raw Xinerama row against both logical and normalized pixel coordinates before using the ordering fallback. Focused wiring assertions, QML lint, Control Center checks, local CodeRabbit review, and the full `make check` pass.

File: config/quickshell/shell.qml [hunk 8]
```diff
@@ -356,29 +378,38 @@ ShellRoot {
...
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply config/quickshell/shell.qml:389]: Fixed in 99f9f70. Profiling found the idle cost was an `_NET_CLIENT_LIST` storm: dwm deleted and re-appended partial client lists on repeated panel ConfigureNotify events, producing thousands of root-property notifications. `updateclientlist()` now builds one atomic list and publishes only when it differs from the cached value; manage/tray paths use the same atomic update. A nested closed-launcher X11 sample with the new dwm measured 0.00% Quickshell CPU over 10 seconds and only the initial client-list event over 5 seconds. The live dwm session was intentionally not restarted.

File: dwm.c [hunk 9]
```diff
@@ -5126,30 +5157,51 @@ getmontagmask(int monnum)
...
-		/* Fallback to first tag if calculation fails */
-		if (!m->tagset[0])
-			m->tagset[0] = m->tagset[1] = 1;
+		if (!fallbacktag)
+			fallbacktag = 1;
+
+		for (s = 0; s < 2; s++) {
+			m->tagset[s] &= montags;
+			if (!m->tagset[s])
+				m->tagset[s] = fallbacktag;
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply dwm.c:5322]: Fixed in 99f9f70. Before monitor tagsets are masked, `reconcilemonitortags()` now walks clients whose tags no longer overlap their current monitor, locates the monitor that owns those original tag bits, detaches/reattaches the client there, preserves the owned tag bits, and refreshes `_NET_WM_DESKTOP`. The focused monitor-tag guard, nested X11 runtime test, clean C build, local CodeRabbit review, and full `make check` pass.

**Comment by ChrisTitusTech** [comment]
@codex review

File: dwm.c [hunk 10]
```diff
@@ -5126,30 +5182,76 @@ getmontagmask(int monnum)
...
+		}
+
+		for (i = 0; i < LENGTH(tags)
+		     && !(m->tagset[m->seltags] & (1 << i)); i++);
+		m->pertag->curtag = i < LENGTH(tags) ? i + 1 : 1;
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply dwm.c]: Fixed in ee4c394. Reconciliation now preserves the active per-tag slot when it remains present, uses slot 0 when the active tagset equals the monitor-owned all-tags mask, and only falls back to the first set bit when the prior slot is no longer valid. The same semantics are applied to `prevtag`. Focused monitor-tag checks, clean build, and full `make check` pass.

File: dwm.c [hunk 11]
```diff
@@ -4518,22 +4518,55 @@ updateclientlist(void)
...
 		for (c = m->clients; c; c = c->next)
-			ewmh_append_client(c->win);
+			clients[index++] = c->win;
 	}
+	count = index;
+
+	if (clientlistcachevalid && count == clientlistcachelen
+	&& (!count || !memcmp(clients, clientlistcache, count * sizeof(Window)))) {
+		free(clients);
+		return;
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply dwm.c:4612]: Fixed in ee4c394. `setclientdesktop()` now increments and publishes the watched `DWM_TAG_UPDATE` root property after every `_NET_WM_DESKTOP` assignment, so occupied-workspace state refreshes even when the atomic `_NET_CLIENT_LIST` cache correctly suppresses an unchanged list. Focused monitor-tag checks and full `make check` pass.

File: config/quickshell/shell.qml [hunk 12]
```diff
@@ -356,29 +378,38 @@ ShellRoot {
...
+            required property var modelData
+
+            screen: modelData
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply config/quickshell/shell.qml:392]: Fixed in ee4c394. Root geometry handling now resizes each external panel with that monitor's recorded `m->bh` instead of the global `bh`, preserving native per-screen panel heights across mixed-DPI reconfiguration. The clean C build, focused guard, and full `make check` pass.

File: dwm.c [hunk 13]
```diff
@@ -5126,30 +5182,76 @@ getmontagmask(int monnum)
...
+			for (owner = mons; owner; owner = owner->next)
+				if (c->tags & getmontagmask(owner->num))
+					break;
+			if (!owner || owner == m)
+				continue;
+
+			detach(c);
+			detachstack(c);
+			c->mon = owner;
+			c->tags &= getmontagmask(owner->num);
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply dwm.c:5292]: Fixed in ee4c394. Floating clients moved to a new tag owner now translate their coordinates relative to the destination monitor, clamp within its work area, and are resized at the translated position after reattachment. Focused monitor-tag checks, clean build, and full `make check` pass.

File: dwm.c [hunk 14]
```diff
@@ -5126,30 +5182,76 @@ getmontagmask(int monnum)
...
+	int i, s;
 	
 	updatemonitorcount();
+
+	for (m = mons; m; m = m->next) {
+		montags = getmontagmask(m->num);
+		for (c = m->clients; c; c = next) {
+			next = c->next;
+			if (c->tags & montags)
+				continue;
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply dwm.c]: Fixed in ee4c394. During topology reconciliation, non-sticky clients that retain any current-monitor overlap are now masked to that monitor's owned tag range and publish their updated desktop assignment. Sticky `TAGMASK` clients retain their documented all-tags behavior. Focused monitor-tag checks and full `make check` pass.

**Comment by ChrisTitusTech** [comment]
@codex review

File: dwm.c [hunk 15]
```diff
@@ -1014,10 +1020,11 @@ configurenotify(XEvent *e)
 						resizeclient(c, m->mx, m->my, m->mw, m->mh);
 				}
 				if (m->barwin)
-					XMoveResizeWindow(dpy, m->barwin, m->wx, m->by, m->ww, bh);
+					XMoveResizeWindow(dpy, m->barwin, m->wx, m->by, m->ww, m->bh);
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply dwm.c:1027]: Fixed in 702fd08. `togglebar()` now resizes the selected external panel with `selmon->bh`, matching the root geometry path and that monitor's reserved work area. Focused monitor-tag checks, clean build, and full `make check` pass.

File: config/quickshell/state/DwmState.qml [hunk 16]
```diff
@@ -59,6 +74,81 @@ Scope {
...
+        const indexes = [];
+        const workspaceCount = root.workspaceNames.length;
+
+        if (workspaceCount === 0) {
+            return indexes;
+        }
+
+        const screenCount = Math.max(1, Quickshell.screens.length);
+        const logicalIndex = Math.min(root.screenIndex(screen), screenCount - 1);
+        const workspacesPerScreen = Math.max(1, Math.floor(workspaceCount / screenCount));
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply config/quickshell/state/DwmState.qml]: Fixed in 702fd08. Workspace partitioning now uses the published `_DWM_MONITOR_DESKTOPS` row count whenever available, falling back to `Quickshell.screens.length` only for older dwm instances without monitor rows. Cloned outputs therefore retain the single logical monitor's full tag range. Focused Control Center/QML checks and full `make check` pass.

**Comment by ChrisTitusTech** [comment]
@codex review

File: config/quickshell/state/DwmState.qml [hunk 17]
```diff
@@ -59,6 +74,82 @@ Scope {
         return root.occupiedWorkspaces.indexOf(index) !== -1;
     }
 
+    function screenIndex(screen) {
+        const pixelRatio = screen && screen.devicePixelRatio > 0
+            ? screen.devicePixelRatio : 1;
+        const pixelX = screen ? Math.round(screen.x * pixelRatio) : 0;
+        const pixelY = screen ? Math.round(screen.y * pixelRatio) : 0;
+        const pixelWidth = screen ? Math.round(screen.width * pixelRatio) : 0;
+        const pixelHeight = screen ? Math.round(screen.height * pixelRatio) : 0;
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply config/quickshell/state/DwmState.qml]: Fixed in fe6e947. `screenIndex()` now preserves Qt's native virtual-desktop `x`/`y` positions while applying `devicePixelRatio` only to the reported screen dimensions. This keeps mixed-DPI geometries in the same coordinate space as `_DWM_MONITOR_DESKTOPS` and avoids the ordering fallback. Focused regression guards, Quickshell-aware lint, the clean C build, and full `make check` pass.

File: dwm.c [hunk 18]
```diff
@@ -1006,6 +1011,7 @@ configurenotify(XEvent *e)
 		sw = ev->width;
 		sh = ev->height;
 		if (updategeom() || dirty) {
+			reconcilemonitortags();
 			drw_resize(drw, sw, bh);
 			updatebars();
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply dwm.c:1018]: Fixed in fe6e947. Root `ConfigureNotify` handling now rebuilds `_NET_CLIENT_LIST` immediately after monitor reconciliation and bar reconstruction, so a removed monitor's panel XID cannot remain cached when its later `DestroyNotify` no longer maps to a live `Monitor`. The focused ordering guard, clean build, and full `make check` pass.

File: config/quickshell/shell.qml [hunk 19]
```diff
@@ -356,29 +378,38 @@ ShellRoot {
...
 
-        state: dwmState
-        clock: clock
-        networkModel: networkModel
-        controlsModel: controlsModel
-        bluetoothModel: bluetoothModel
-        controlCenterModel: controlCenterModel
-        powerMenuModel: powerMenuModel
+        model: Quickshell.screens
+
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply config/quickshell/shell.qml:388]: Rechecked on the current head without restarting or disturbing the live Fedora X11 session: the managed Quickshell process with two active 2560x1440 panels measured 3.00%, 2.90%, and 3.00% of one CPU core across three consecutive 10-second closed-idle samples. The isolated System Health Xvfb 5% gate also passes, and Settings reports 0.00% closed-idle CPU. The earlier single 5.7% observation is not reproducible, so no speculative performance change is warranted.

**Comment by ChrisTitusTech** [comment]
@coderabbitai review

File: dwm.c [hunk 20]
```diff
@@ -1006,18 +1011,21 @@ configurenotify(XEvent *e)
...
+			updateclientlist();
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply dwm.c:1020]: Fixed in 13ef114. Root geometry handling now rediscovers every mapped dock panel by its live geometry before rebuilding `_NET_CLIENT_LIST`, which rebinds a surviving panel even when its former `Monitor` node was freed. Later dock `ConfigureNotify` events also re-evaluate the destination monitor; a cross-monitor move clears and rearranges the previous owner before rebinding, while an intentionally hidden offscreen panel retains its association. Focused hotplug guards, raw GCC build, local CodeRabbit rerun with zero findings, and full `make check` pass. A local Clang rerun was unavailable because `clang` is not installed; GitHub's Clang job is running for this commit.

**Comment by ChrisTitusTech** [comment]
@coderabbitai review

File: dwm.c [hunk 21]
```diff
@@ -1006,29 +1013,44 @@ configurenotify(XEvent *e)
...
-				arrange(m);
-				XRaiseWindow(dpy, ev->window);
-				updateclientlist();
-				break;
-			}
+	} else if (XGetWindowAttributes(dpy, ev->window, &wa)) {
+		for (oldm = mons; oldm && oldm->barwin != ev->window;
+		     oldm = oldm->next);
+		if (getwinatomprop(ev->window, netatom[NetWMWindowType])
+		    == netatom[NetWMWindowTypeDock]) {
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply dwm.c]: Fixed in 726a823. Both the dock ConfigureNotify rebind path and scanaltbars() now apply isaltbar() before updatealtbar(), preserving the existing class-based panel recognition while rejecting vertical generic docks. The focused monitor guard, clean GCC build, ShellCheck/shfmt, and full make check all pass.

File: dwm.c [hunk 22]
```diff
@@ -2785,6 +2807,38 @@ scan(void)
...
+	for (i = 0; i < num; i++) {
+		if (!XGetWindowAttributes(dpy, wins[i], &wa)
+		    || wa.override_redirect || wa.map_state != IsViewable
+		    || getwinatomprop(wins[i], netatom[NetWMWindowType])
+		        != netatom[NetWMWindowTypeDock])
+			continue;
+
+		m = recttomon(wa.x, wa.y, wa.width, wa.height);
+		if (!m || INTERSECT(wa.x, wa.y, wa.width, wa.height, m) <= 0)
+			continue;
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply dwm.c:2847]: Fixed in 726a823. scanaltbars() now snapshots each live monitor's known panel XID before rebuilding registrations. A mapped panel with no current on-screen intersection falls back to that prior owner, so togglebar()-hidden panels remain registered; a failed XQueryTree also leaves existing registrations intact. The focused monitor guard, clean GCC build, and full make check pass.

File: dwm.c [hunk 23]
```diff
@@ -1006,29 +1013,44 @@ configurenotify(XEvent *e)
...
-		for (m = mons; m; m = m->next) {
-			if (m->barwin == ev->window &&
-			    XGetWindowAttributes(dpy, ev->window, &wa) &&
-			    updatealtbar(m, ev->window, &wa)) {
-				arrange(m);
-				XRaiseWindow(dpy, ev->window);
-				updateclientlist();
-				break;
-			}
+	} else if (XGetWindowAttributes(dpy, ev->window, &wa)) {
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply dwm.c]: Fixed in 726a823. The non-root ConfigureNotify path now rejects wintoclient() matches before XGetWindowAttributes() or isaltbar() performs any synchronous X server queries. Only untracked windows remain eligible for panel classification and rebind. The ordering is covered by the focused source guard; the clean GCC build and full make check pass.

File: dwm.c [hunk 24]
```diff
@@ -5126,30 +5234,101 @@ getmontagmask(int monnum)
...
+				c->x = MAX(owner->wx,
+					MIN(c->x, owner->wx + owner->ww - WIDTH(c)));
+				c->y = MAX(owner->wy,
+					MIN(c->y, owner->wy + owner->wh - HEIGHT(c)));
+			}
+			c->mon = owner;
+			c->tags &= getmontagmask(owner->num);
+			setclientdesktop(c);
+			attachbottom(c);
+			attachstack(c);
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply dwm.c:5295]: Fixed in 726a823. Before migration, reconcilemonitortags() records whether the client was selected and whether its source monitor was active. After attaching it to the destination, it restores owner->sel and moves selmon for the previously focused client, allowing the geometry handler's existing focus(NULL) to retain active focus. The focused guard, clean GCC build, and full make check pass.

**Comment by ChrisTitusTech** [comment]
@codex review

@coderabbitai review

File: dwm.c [hunk 25]
```diff
@@ -5126,30 +5246,107 @@ getmontagmask(int monnum)
...
+			if (wasselected)
+				owner->sel = c;
```
[the remark this answers is not in the document]
Comment (ChrisTitusTech) [reply dwm.c:5297]: No code change is needed here. Before c->mon is reassigned, detachstack(c) runs against the source monitor. Its existing selected-client branch explicitly checks c == c->mon->sel and replaces that source selection with the first visible client remaining in c->mon->stack, or NULL when none remains. Only after that source pointer is cleared does reconciliation set c->mon = owner, attach it to the destination stack, and restore owner->sel. The focused source guard and full make check pass on 726a823, so the described dangling source selection cannot occur with this ordering.
